File: problem060.jl
using Primes
function compatible(a::Int, b::Int)
	r = b
	while r > 0
		a *= 10
		r = div(r, 10)
	end
	isprime(a + b)
end

function generate(n::Int, pri::Array{Int64,1})
	if n == 1
		for i in pri
			produce([i])
		end
	else
		g = @task generate(n - 1, pri)
		while true
			t = consume(g)
			i = 1
			p = 2
			while p < t[1]
				valid = true
				for j in t
					if !compatible(j, p) || !compatible(p, j)
						valid = false
						break
					end
				end
				if valid
					produce(vcat(p, t))
				end
				i += 1
				p = pri[i]
			end
		end
	end
end

function findNumbers()
	pri = primes(10^6)
	g = @task generate(5, pri)
	vals = consume(g)
	println(vals, ", sum: ", sum(vals))
end
t = time()
findNumbers()
t = time() - t
print(t)

File: problem060.py
import math
def primes(n):
	t = [0] * (n + 1)
	t[0] = 1
	t[1] = 1
	for i in range(2, n//2 + 1):
		if t[i] == 0:
			j = i + i
			while j <= n:
				t[j] = 1
				j += i
	pri = [2]
	for i in range(3, n):
		if t[i] == 0:
			pri.append(i)
	return pri
primeList = primes(1000000)
def isPrime(n):
	if n < 2:
		return False
	l = math.sqrt(n)
	i = 1
	p = 2
	while p <= l:
		if n % p == 0:
			return False
		p = primeList[i]
		i += 1
	return True
def compatible(a, b):
	r = b
	while r > 0:
		a *= 10
		r = r // 10
	return isPrime(a + b)
def generate(n, pri):
	if n == 1:
		for i in pri:
			yield [i]
	else:
		g = generate(n - 1, pri)
		while True:
			t = g.__next__()
			i = 0
			p = 2
			while p < t[0]:
				valid = True
				for j in t:
					if (not compatible(j, p)) or (not compatible(p, j)):
						valid = False
						break
				if valid:
					# if n > 3:
					# 	print("yielding :", [p] + t)
					yield [p] + t
				i += 1
				p = pri[i]

# # print(compatible(3,7))
# print(primeList[0:32])
g = generate(5, primeList)
res = g.__next__()
print(res)
print(sum(res))


